src/main/scala/reporting/DomainInterpreter.scala

```diff
@@ -0,0 +1,51 @@
+package viper.silicon.reporting
+import viper.silicon.state.terms._
+import viper.silicon.state.Identifier
+
+/**
+  * the most Generic form of interpreter
+  *  @param	E : input type
+  *  @param I : info object type
+  *  @param	T : result type
+  */
+trait AbstractInterpreter[E, I, T] { 
+	def interpret(entry: E, info: I): T
+}
+trait ModelInterpreter[T, I] extends AbstractInterpreter[ExtractedModelEntry, I, T]
+trait DomainInterpreter[T, I] extends AbstractInterpreter[DomainValueEntry, I, T]
+
+case class IdentityInterpreter() extends ModelInterpreter[ExtractedModelEntry, Any] {
+	def interpret(entry: ExtractedModelEntry, anything: Any): ExtractedModelEntry = entry
+}
+
+/**
+  * Simple domain interpreter resolves entries in domains recursively otherwise it just returns the value of the entry
+  * does not resolve domain values stored in sequences or Refs
+  * @param c Converter that has all necessary infos about the counterexample domains
+  */
+case class GenericDomainInterpreter(c: Converter) extends ModelInterpreter[ExtractedModelEntry, Seq[ExtractedModelEntry]] {
+	private val domains: Seq[DomainEntry] = c.domains
+	private val functions: Seq[ExtractedFunction] = c.non_domain_functions
+	def interpret(entry: ExtractedModelEntry,visited: Seq[ExtractedModelEntry]): ExtractedModelEntry = {
+		if (visited.contains(entry)) entry
+		else entry match {
+			case DomainValueEntry(dom, v) =>
+				val relevantDomain = domains.filter(x => x.valueName == dom)
+				val domainFunctions = relevantDomain.flatMap(_.functions)
+				val allFunctions = functions ++ domainFunctions
+				val relevantFunctions = allFunctions.filter(x => x.argtypes.length == 1 && x.argtypes.head == sorts.UserSort(Identifier.apply(dom)))
```

Comment:
I don't see explicit `x.apply(y)` call very often; just `x(y)` appears to be more common.